### DISCUSSION TITLE
Multi-line with kubernetes source?

### DISCUSSION THREAD

**Discussion by danthegoodman1** [vectordotdev/vector#13080] [Q&A] [answered]
Things like Go, when panicking, will print multiple lines out.

Is there any way to add multi-line support to kubernetes? I see that it is available with the file source, but I don't see any settings within the k8s source that would allow multi-line

**Comment by jszwedko** [comment, marked as answer]
Hi @danthegoodman1 !

We've actually been moving away from having multi-line support at the source level and instead suggesting users use the `reduce` transform which can be used with any source to combine multi-line logs. That should work for you, but let us know if you have any issues!

**Reply by danthegoodman1** [reply]
That seems wise! Thanks

**Comment by jblang** [comment]
Doesn't using reduce this early in the pipeline limit concurrency? I don't have scientific measurements but I was trying to use reduce for multiline coalescing in conjunction with the file source, in anticipation of eventually using the kubernetes source, and when I had a large backlog of files to process, the reduce appeared to cause a bottleneck that was drastically improved by using the native multiline features in the file source.

**Reply by jszwedko** [reply]
Ah, yeah, that's a good note @jblang . `reduce` is single-threaded at the moment. We do hope to improve that in the future by partitioning events along `group_by` fields to run multiple `reduce` transforms concurrently.

**Reply by danthegoodman1** [reply]
Was about to mention this, crushing the CPU with the amount of logs we generate per host

**Reply by danthegoodman1** [reply]
@jszwedko Is there a way we could use the file source instead of the kubernetes source? I imagine that the k8s source is just looking at the file system anyway :P

**Reply by jblang** [reply]
The solution we landed on was to use Fluent Bit's [tail input](https://docs.fluentbit.io/manual/pipeline/inputs/tail) with a [multiline parser](https://docs.fluentbit.io/manual/administration/configuring-fluent-bit/multiline-parsing) in conjunction with their [kubernetes filter](https://docs.fluentbit.io/manual/pipeline/filters/kubernetes) to enrich the logs with k8s metadata. It then forwards the events to Vector using Vector's fluent source, and Vector then further parses and enriches the events and forwards them to Splunk.  We are using Fluent Bit tail instead of Vector's file source for a few reasons:

1) Vector requires a newer version of Kubernetes than we are currently using.
2) Vector's Kubernetes source doesn't support multiline
3) Fluentbit [has a feature](https://docs.fluentbit.io/manual/pipeline/filters/kubernetes#optional-feature-using-kubelet-to-get-metadata) to query k8s metadata from the local kubelet instead of the API server which alleviates a problem we had with filebeat where the k8s API server gets overwhelmed and causes events to get delayed or dropped.

In general, I think Vector is much better designed than Fluent Bit and would prefer to use it for everything, but until these issues are resolved, this was the best solution we could find.  If Vector implemented k8s enrichment as a transform rather than a source, then we could use the Vector file source with mulitline enabled instead of Fluent Bit, but we'd probably still need the ability to query the kubelet rather than the API server and support for our k8s versions (1.18 is the lowest we need to support).  

If you only need the pod name, namespace, container name, and container ID, you could use the file source and parse these from the filename using regex, but if you need the k8s pod id, labels, or annotations, there is no choice but to hit the k8s API.

**Reply by jszwedko** [reply]
Indeed, it just wraps the same logic the `file` source uses. Uou can use the `file` source, you just won't get the enrichment of events from k8s metadata (like pod name, namespace, etc.). To do that, you can just point the `file` source at the directory containing logs (usually `/var/logs/pods/...`).

**Reply by jszwedko** [reply]
Thanks for those notes @jblang ! We have been thinking about separating out the k8s enrichment to a separate transform too. This is even more motivation to do that 😄

**Reply by spencergilbert** [reply]
> If Vector implemented k8s enrichment as a transform rather than a source, then we could use the Vector file source with mulitline enabled instead of Fluent Bit, but we'd probably still need the ability to query the kubelet rather than the API server.

Noting that I'd also like to do these two things @jblang  🙂

**Reply by danthegoodman1** [reply]
Yeah sucks to have to use 2 different logging solutions though. If I were doing this I probably would just run with fluentbit to reduce complexity.

Would love to see k8s enrichment as a transform!

**Reply by danthegoodman1** [reply]
@jblang how do you set up fluentbit to expose the logs over TCP, is that a sink?

**Reply by spencergilbert** [reply]
@danthegoodman1 - https://vector.dev/docs/reference/configuration/sources/fluent/#fluent-bit-configuration

**Reply by danthegoodman1** [reply]
thanks!

**Reply by jblang** [reply]
> Yeah sucks to have to use 2 different logging solutions though. If I were doing this I probably would just run with fluentbit to reduce complexity.

Agreed, but I've found implementing complex parsers in Fluent Bit to be very painful and Vector brings enough to the table that it's worth using both for me.  We try to do as little as possible in Fluent Bit and then hand off to Vector as soon as possible, so we can keep most of our parsing logic there.  Also, since Fluent Bit has to be deployed on every node, but we only have two Vector aggregators per region, rolling out changes is much faster when we only have two Vector pods to restart. Even if we moved to Vector entirely, I'd probably still keep this topology and division of labor.

**Reply by jblang** [reply]
> @jblang how do you set up fluentbit to expose the logs over TCP, is that a sink?

The link Spencer shared is exactly how we are doing it.  You may also find the corresponding Fluent Bit docs for the [forward output](https://docs.fluentbit.io/manual/pipeline/outputs/forward) helpful.

**Reply by jszwedko** [reply]
@jblang just curious, what Kubernetes version are you running that Vector doesn't support?

**Reply by jblang** [reply]
> @jblang just curious, what Kubernetes version are you running that Vector doesn't support?

We have multiple k8s clusters we need to use with Vector.  Depending on the cluster, we have 1.18, 1.19 and 1.20.

**Reply by jszwedko** [reply]
Ah, gotcha, thanks @jblang . We do test down to 1.19 (https://github.com/vectordotdev/vector/blob/6584d58b5b08b887e2a29c840e751c6283e2d1b5/.github/workflows/k8s_e2e.yml#L107-L113). I would have thought 1.18 would work too but it sounds like not :)

**Reply by jblang** [reply]
To be honest I haven't personally verified whether it works with 1.18 or not.  I was going by the note in the docs stating that 1.19 is required. I do know that several APIs changed from v1alpha1 to v1beta1 around this time and we have had problems talking to our 1.18 clusters with kubectl newer than 1.19.  Anyway we have plans to upgrade our 1.18 clusters soon so hopefully this will not be an issue for us for too long.

**Reply by spencergilbert** [reply]
I'm _pretty_ sure we should work with v1.18 - I'll dig for evidence and make the change if it's the case, but I think the "requirement" is probably still down towards 1.12/1.13 since I don't think any of the API's we call have changed since then.

**Reply by jblang** [reply]
> `reduce` is single-threaded at the moment. We do hope to improve that in the future by partitioning events along `group_by` fields to run multiple `reduce` transforms concurrently.

I suspected that this was the case based on the CPU utilization I was seeing.  It would be great to have it parallelize by the groups in the future.
